midend/hsIndexSimplify.h

```diff
@@ -0,0 +1,96 @@
+#ifndef _MIDEND_HSINDEXSIMPLIFY_H_
+#define _MIDEND_HSINDEXSIMPLIFY_H_
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+
+namespace P4 {
+
+typedef std::map<cstring, const IR::PathExpression*> generatedVariablesMap;
+
+/// This class finds innermost header stack with non-concrete index.
+/// For each found innermost header stack it generates new local variable and
+/// adds it to the corresponded local definitions.
+class HSIndexFinder : public Inspector {
+    friend class HSIndexTransform;
+    friend class HSIndexSimplifier;
+    IR::IndexedVector<IR::Declaration>* locals;
+    ReferenceMap* refMap;
+    TypeMap* typeMap;
+    const IR::ArrayIndex* arrayIndex;
+    const IR::PathExpression* newVariable;
+    generatedVariablesMap* generatedVariables;
+    std::set<cstring> storedMember;
+    std::list<IR::Member*> dependedMembers;
+
+ public:
+    HSIndexFinder(IR::IndexedVector<IR::Declaration>* locals, ReferenceMap* refMap,
+                  TypeMap* typeMap, generatedVariablesMap* generatedVariables)
+        : locals(locals),
+          refMap(refMap),
+          typeMap(typeMap),
+          arrayIndex(nullptr),
+          newVariable(nullptr),
+          generatedVariables(generatedVariables) {}
+    void postorder(const IR::ArrayIndex* curArrayIndex) override;
+
+ protected:
+    void addNewVariable();
+};
+
+/// This class substitutes index of a header stack in all occurence of found header stack.
+class HSIndexTransform : public Transform {
+    friend class HSIndexSimplifier;
+    int index;
+    HSIndexFinder& hsIndexFinder;
+
+ public:
+    HSIndexTransform(HSIndexFinder& finder, int index) : index(index), hsIndexFinder(finder) {}
+    const IR::Node* postorder(IR::ArrayIndex* curArrayIndex) override;
+};
+
+/// This class eliminates all non-concrete indexes of the header stacks in the controls.
+/// It generates new variables for all expressions in the header stacks indexes and
+/// checks their values for substitution of concrete values.
+/// Restriction : in/out parameters should be replaced by correspondent assignments.
```

Comment:
you have another precondition, which is that array variable names are really unique within a scope (there is no aliasing).
I think you should list that. This is ensured by prior passes. 
So you won't have `{ H[2] a; { H[3] a; ... } }`